### DISCUSSION TITLE
make pyright auto-detect zip files like cpython

### DISCUSSION THREAD

**Discussion by DanielPenning** [microsoft/pyright#9192] [General]
We employ zip-files that contain python packages. For usability reasons these zip files do not have the extension `.zip`, but a product-specific one like `xyz.product`. When I add the full path to PYTHONPATH, cpython (tested with 3.10.6) is able to automatically detect that this is a zip file and makes the module available for import.  So cpython seems to perform some auto-magic to detect that this file is a zip-file and handles it properly.

**Problem**
However pyright seems to handle this differently. I observe this indirectly through VS Code / Pylance.
With the PYTHONPATH adjustment above, this import is not resolved. WhenI configure an extra analysis path:
` "python.analysis.extraPaths": [ 'c:/Users/User/configdir/xyz.project"]`
I see an error message that points to the reason within pyright.

```
[info] [Error - 10:27:50 AM] (3316) Indexing Failed: "Exception: ENOTDIR: not a directory, scandir 'c:\\Users\\User\configdir\\xyz.project' in Error: ENOTDIR: not a directory, scandir 'c:\\Users\\User\\configdir\\xyz.project'
    at Object.readdirSync (node:fs:1509:26)
    at Object.t.readdirSync (node:electron/js2c/node_init:2:11418)
    at A.readdirSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\vendor.bundle.js:2:779302)
    at makeCallSync.subPath (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\vendor.bundle.js:2:826524)
    at w.makeCallSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\vendor.bundle.js:2:828316)
    at w.readdirSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\vendor.bundle.js:2:826492)
    at u.readdirSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\vendor.bundle.js:2:786461)
    at S.readdirSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\vendor.bundle.js:2:786461)
    at _0x5e80d2.readdirEntriesSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\pyright.bundle.js:1:1215239)
    at l.readdirEntriesSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\pyright.bundle.js:1:1599045)
    at _0x421d1f.readdirEntriesSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:495729)
    at _0x72b57c.readdirEntriesSync (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:900009)
    at _0x350251.scan (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:894537)
    at c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:1246523
    at t.LogTracker.log (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\pyright.bundle.js:1:1197428)
    at _0x507b02.scanPackages (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:1245766)
    at c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:1261028
    at t.run (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\pyright.bundle.js:1:1095351)
    at MessagePort.<anonymous> (c:\\Users\\User\\.vscode\\extensions\\ms-python.vscode-pylance-2024.10.1\\dist\\server.bundle.js:1:1260415)
    at MessagePort.[nodejs.internal.kHybridDispatch] (node:internal/event_target:820:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)"
```

When I rename to `xyz.zip` everything works fine. 

**Discussion for change**
I tracked down this problem to `packages\pyright-internal\src\common\realFileSystem.ts`
There is indeed already magic implemented to auto-detect ".egg" files as zip. 

```
// "Magic" values for the zip file type. https://en.wikipedia.org/wiki/List_of_file_signatures
const zipMagic = [
    Buffer.from([0x50, 0x4b, 0x03, 0x04]),
    Buffer.from([0x50, 0x4b, 0x05, 0x06]),
    Buffer.from([0x50, 0x4b, 0x07, 0x08]),
];
```

This is exactly what I need in a generic way (to support our `.project` files).
It would be great if this method could just be applied to files in general, exactly like cpython does for PYTHONPATH.

`realFileSystem.ts` does some for ZipOpenFS to get the `.egg` files handled properly. The code has an interesting comment from 3 years ago:

```
// Patch fslib's ZipOpenFS to also consider .egg files to be .zip files.
//
// For now, override findZip (even though it's private), with the intent
// to upstream a change to allow overriding getArchivePart or add some
// other mechanism to support more extensions as zips (or, to remove this
// hack in favor of a full ZipOpenFS fork).
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
//@ts-expect-error
class EggZipOpenFS extends ZipOpenFS {
```

Is there interest to get this implemented? Would a Pull Request be considered? If yes, is there an approach that would be preferred?

**Comment by erictraut** [comment]
The error message you cited above is from pylance's indexer. This functionality isn't included in pyright but is instead layered on top of pyright in the closed-source pylance project. If you'd like the pylance indexer to work with zip files, you'd need to file an enhancement request with the pylance team in the [pylance-release](https://github.com/microsoft/pylance-release) project.

It's not clear from your post whether the core pyright functionality works with your use case. Could you confirm? If this also affects core pyright functionality, we could consider adding this support. However, this is very specific to your use case, so I'd want to make sure that any such change doesn't negatively impact pyright functionality or performance for other pyright users. The import resolution paths in pyright are performance sensitive, so I would be concerned about needing to open arbitrary files and reading their contents to determine whether they might contain zipped files.

**Comment by DanielPenning** [comment]
Hi @erictraut 

thanks for getting back to me so quickly. I should have reproduced this with pyright itself from the beginning. Sorry for this lazyness on my part. Here are the steps to show a mismatch in PYTHONPATH handling between cpython and pyright (tested versions are cpython 3.10.6, pyright 1.1.384).

Directory structure
- ./use_xyz.py
- ./xyz.zip with one file in it: foo.py

use_xyz.py
```
import foo

foo.bar()
```

foo.py
```
def bar():
    print("foo.py bar function called")
```

Adding `xyz.zip` to PYTHONPATH allows cpython to find it and pyright can include it properly:
```
PS C:\src\code-completion> pwsh -Command { $env:PYTHONPATH="xyz.zip"; python use_xyz.py }    
foo.py bar function called
PS C:\src\code-completion> pwsh -Command { $env:PYTHONPATH="xyz.zip"; pyright use_xyz.py }   
0 errors, 0 warnings, 0 informations
```

Renaming `xyz.zip` to `xyz.project` still allows cpython to import it but pyright does not.

```
PS C:\src\code-completion> pwsh -Command { $env:PYTHONPATH="xyz.project"; python use_xyz.py }
foo.py bar function called
PS C:\src\code-completion> pwsh -Command { $env:PYTHONPATH="xyz.project"; pyright use_xyz.py }
C:\src\code-completion\use_xyz.py:1:8 - error: Import "foo" could not be resolved (reportMissingImports)
1 error, 0 warnings, 0 informations
```

Regarding performance: For all entries in PYTHONPATH that are _not_ directories, they need to be checked for zip-automagic. As I said above the code to detect that is already inside `packages\pyright-internal\src\common\realFileSystem.ts` but only active for `.egg` extensions.
Let me know if these steps for reproduction help. Again sorry for not putting it this way from the beginning.

**Reply by erictraut** [reply]
As you've pointed out, pyright's handling for zip files is in the `realFileSystem.ts` module, specifically in the `readdirEntriesSync` method. This method is on the hot path (i.e. it's performance critical) for import resolution. The current logic performs a lightweight check specifically for files with an ".egg", ".zip" or ".jar" file extension. It does not attempt to open files with other file extensions. You're proposing that pyright open and examine all files in this code path to see if they might contain zip contents. 

I added some temporary logging to better understand the potential impact of such a change. I found that it would result in an average of 2,500 additional file opens during a typical type checking run. I saw as many as 45,000 additional file opens for larger and more complex projects. This translates to measurable performance degradations for pyright users, so I don't think that would be a good tradeoff unless you can show that your use case is relatively common among pyright users.

Is there a reason you use a ".project" file extension rather than one of the standard file extensions for zip files? Is it feasible for you to switch?

**Reply by DanielPenning** [reply]
> You're proposing that pyright open and examine all files in this code path to see if they might contain zip contents.

No. I only propose to check files (!) that are explicitly mentioned in PYTHONPATH list.
I do not propose to check files that are found in directories within PYTHONPATH (I think only this would cause the performance penalty you found in your logging).

My proposal should add next to zero overhead because it is rare that files are in the PYTHONPATH list and if yes, they are almost certain a .zip file.  Or do I miss something here?
Maybe the method I mentioned above in realFileSystem.ts is not suitable for this and it must be handled higher-up. 

As I mentioned, I am willing to let somebody take a look to implement this once the approach is agreed upon.
Unfortunately switching to another extension is not possible for us. And I think the fact thy cpython itself already supports this is an argument to handle it the same way in pyright / static analyzers in general.

Again, thanks for taking your time to look into this.

**Reply by erictraut** [reply]
There are many behaviors implemented in CPython that are not supported by static type checkers, so that's not a compelling argument in itself.

Can you tell me more about who "we" are in this case? I don't know anything about your use case other than what you've told me above, so I can't tell if this affects a million users, a dozen users, or something in between.

**Reply by DanielPenning** [reply]
Sure ;) We are a startup that builds a new kind of test system for embedded devices.
Our solution is based on VS Code and Python. In the user number, we are currently more on the lower end of your estimate.
However If we can agree that performance is not negatively affected (as I have argued above) and we provide a simple PR for this, would you be willing to take a look at it? 
If there are no arguments against it, I do not see a compelling argument to _not_ match cpython behaviors more closely?

**Reply by erictraut** [reply]
The original `realFileSystem.ts` module and the zip support was contributed to pyright by the Microsoft pylance team. I took a closer look at the code, and it makes some tenuous assumptions. I also don't see any test cases for it, which increases the chances of regressions if we were to modify it. I think that this change is more involved than you appreciate.

If we were to add the suggested functionality, I think it would be best for the pylance team to make the change to ensure that it's done without regressions (either functional or performance). I recommend filing an enhancement request in the [pylance-release](https://github.com/microsoft/pylance-release/issues) issue tracker. The pylance team typically prioritizes their work based on community feedback (thumbs-up reactions).
